updatehub-sdk/src/listener.rs

```diff
@@ -0,0 +1,112 @@
+// Copyright (C) 2018, 2019, 2020 O.S. Systems Sofware LTDA
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{Error, Result};
+use async_std::{
+    io::BufReader,
+    os::unix::net::{UnixListener, UnixStream},
+    prelude::*,
+};
+use log::warn;
+use std::{env, fs, path::Path, pin::Pin};
+
+const SDK_TRIGGER_FILENAME: &str =
+    "/usr/share/updatehub/state-change-callbacks.d/10-updatehub-sdk-statechange-trigger";
+const SOCKET_PATH: &str = "/run/updatehub-statechange.sock";
+
+type CallbackFn = dyn Fn(Handler) -> Pin<Box<dyn Future<Output = Result<()>>>>;
+
+#[derive(Default)]
+pub struct StateChangeListener {
+    download_callbacks: Vec<Box<CallbackFn>>,
+    install_callbacks: Vec<Box<CallbackFn>>,
+    reboot_callbacks: Vec<Box<CallbackFn>>,
+    error_callbacks: Vec<Box<CallbackFn>>,
+}
+
+pub enum State {
+    Download,
+    Install,
+    Reboot,
+    Error,
+}
+
+pub struct Handler {
+    stream: UnixStream,
+}
+
+impl Handler {
+    // Cancels the current action on the agent.
+    pub async fn cancel(&mut self) -> Result<()> {
+        self.stream.write_all(b"cancel").await.map_err(Error::Io)
+    }
+
+    // Tell the agent to proceed with the transition.
+    pub async fn proceed(&self) -> Result<()> {
+        // No message need to be sent to the connection in order to the
+        // agent to proceed handling the current state.
+        Ok(())
+    }
+}
+
+impl StateChangeListener {
```

Comment:
We should have a `new` **inline** constructor method so we don't really only `Default` trait for building the type.